### PR TITLE
(WIP) Curtail the power of partial_apply

### PR DESCRIPTION
**Pull request by jckarter**
Patches working toward reducing the power of partial_apply and making capture behavior more explicit in SIL.

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please smoke test

File: include/swift/Reflection/TypeRef.h [hunk 1]
```diff
@@ -733,25 +734,32 @@ class UnmanagedStorageTypeRef final : public ReferenceStorageTypeRef {
...
+  static TypeRefID Profile(const TypeRef *BoxedType, bool immutable) {
     TypeRefID ID;
     ID.addPointer(BoxedType);
+    ID.addInteger((uint32_t)immutable);
     return ID;
   }
 public:
-  SILBoxTypeRef(const TypeRef *BoxedType)
-    : TypeRef(TypeRefKind::SILBox), BoxedType(BoxedType) {}
+  SILBoxTypeRef(const TypeRef *BoxedType, bool immutable)
```
Comment (slavapestov) [inline include/swift/Reflection/TypeRef.h]: Unless immutable boxes will get a different mangling you probably don't need to track this in the TypeRef

File: lib/SIL/SILType.cpp [hunk 2]
```diff
@@ -319,7 +319,8 @@ SILType SILType::getEnumElementType(EnumElementDecl *elt, SILModule &M) const {
   // If the case is indirect, then the payload is boxed.
   if (elt->isIndirect() || elt->getParentEnum()->isIndirect())
     loweredTy = SILType::getPrimitiveObjectType(
-      SILBoxType::get(loweredTy.getSwiftRValueType()));
+      SILBoxType::get(loweredTy.getSwiftRValueType(),
+                      /*immutable*/ false));
```
Comment (slavapestov) [inline lib/SIL/SILType.cpp]: Indirect payloads really are immutable though, aren't they? We have no way to refer to them as L values.

Comment (jckarter) [inline lib/SIL/SILType.cpp]: Not today, but that's something we may want to add later.

**Comment by milseman** [comment]
@jckarter do you still want this PR open?

**Comment by jckarter** [comment]
I want to make sure I don't lose it since I'll need some code out of it soon. I'll close it later.

**Comment by tkremenek** [comment]
This was opened in 2016.  Is this still relevant?

**Comment by jckarter** [comment]
I can close it. There are some functions I could salvage out of this if we push further on eliminating partial_apply in the future.

**Pull request closed without merging**